### DISCUSSION TITLE
How to make iOS / MacOS hybrid apps for Big Sur?

### DISCUSSION THREAD

**Discussion by wasserworks** [airsdk/Adobe-Runtime-Support#773] [Q&A]
Air - AIR 33.1.1.444
iMac (Retina 4K, 21.5-inch, 2017)
MacOS Big Sur - 11.2.3
iPad - iOS 4.4.2

Hi, it seems that Big Sur's new hybrid iOS / Mac app will only work on the new silicon Macs, which will make it hard to test. 

Either way, can anyone tell me if the latest Air version supports this? If so, what are the steps to 'allow' the iOS app to work on Mac? 

Anything in the app XML file, or is it done on AppStore Connect?

Thanks,
Matt

**Comment by wasserworks** [comment]
It appears that Macs with the new M1 chip will automatically run all iPad Apps, as the new Appstore in iOS 14 says this under 'Compatibility'

But, does anyone know how the iPad apps display on these Macs? If it scales the full screen of an iPad this would be fine, but if different that would be need more work, and without an M1 Mac I won't be able to test it. 

Matt

**Reply by ajwfrost** [reply]
Yes they do run but there is an open issue with scaling that we need to look into.. (it appears to assume the display is portrait...)

**Reply by wasserworks** [reply]
Ok, thanks Andrew. Please let us know if you solve it somewhere on this forum, or on the Release notes. 

Thanks for new version - works great on iOS 14 and installed on my iPad with Big Sur. Keep up the hard work!

Matt

**Reply by FliplineStudios** [reply]
Oh interesting, I had wondered about that too -- when Apple added the compatibility they actually disabled it for most of our iOS apps because of "display issues", where the landscape apps had what looked like the right 1/3 of the screen covered up with a black box.  That must be where it was making them all act like portrait, while our apps are all landscape-only.  Would be great to have that fixed sometime in the future!

**Reply by wasserworks** [reply]
Hi Andrew / Flipline,
Just checking if the latest Air version fixed this rotation issue on M1 Macs?

Not sure if this Bug fix in .476 version includes this? 

**"AIR-3226: AIR SDK for MacOS to support M1 i.e. universal binaries"**

If so, would be great to hear what the scaling looks like. Is there an option to make it full screen?

**Reply by wasserworks** [reply]
Can anyone please tell me if the iPad app on M1 Mac display has been fixed in a recent Air version yet?

**Reply by FliplineStudios** [reply]
Just checking to see if there's been any update on this, and if anyone knows if the newer versions of AIR have fixed this issue with M1 Macs displaying iOS apps correctly?  We had disabled the "Make this app available on Mac" option for our existing iOS apps since they previously had that portrait display issue, and we've had some customers asking us if we'll ever be supporting that option in an update.  Just curious if it was safe to use for newer AIR builds, or if it's on hold for now...

**Reply by ajwfrost** [reply]
We've not looked at this one yet I'm afraid.. I'll get someone to investigate it to see if it's going to be a quick/easy one.
thanks

**Reply by wasserworks** [reply]
Yes, can you please look into this asap Andrew. At least if you can find out if it's a big or simple job.

I've been waiting for some support to be made for Desktop for literally years now! 

Thanks,
Matt

**Reply by ajwfrost** [reply]
Hi @wasserworks / @FliplineStudios  - just wondering if you have an IPA file from AIR that can even be installed on a mac M1 machine? We're coming up against some slight code-signature differences at the moment which appear to be blocking this. When you double-click an IPA it should "install" to the Applications folder, and the mac is adding some wrapper information. Trying to work around this (i.e. just by using the .app from within a normal .ipa file) and we're coming up against macOS security every time we try something!

We're going to have a quick look at the signature of a "normal" iOS app that works, vs one created from AIR (with the same certs and provisioning profile..) to try to see if we can resolve this and get the IPA to install, but if you'd got an IPA that already installs we may be able to take a look at what issues there are...

thanks

**Reply by wasserworks** [reply]
Thanks for looking into it! I don't have an M1 mac, so I'm in the dark here. I was hoping that when you have resolved it I would try out one of my apps on a friend's M1 laptop! Maybe @FliplineStudios has one?

**Reply by ajwfrost** [reply]
We've just found that the reason our AIR-generated IPA file didn't install is due to code signature problems, where ADT was using the native `codesign` app to sign the IPA file. When we downgraded to Java 11, and did the IPA signing internally within ADT, the IPA file then installed to the M1 machine and it ran. Didn't do much, perhaps we need to use a better test case...

So we can also check why the `codesign` request is causing a problem with this, it seems that the 'get-task-allow' value may be different for some reason...

Anyway: if anyone has an IPA file that's not working properly on macOS M1, please let me know... I'll try to dig out an earlier test case that was showing the resizing problem...

**Reply by ajwfrost** [reply]
FYI the challenge appears to be with the requested device orientation. Even if an app is set up requesting a landscape mode display, it's appearing on an M1 macOS machine in portrait mode. If it goes full-screen then it just scales up the app without adjusting the aspect ratio so we're seeing large black bars to left/right of the AIR display. (That might be partly due to the requested display mode, we'll need to check...)

Anyway -> will see what we can do to fix these issues and to ensure that the generate IPAs have the right entitlements...

thanks

**Reply by wasserworks** [reply]
Sounds very technical! Ok, good luck with it. Desktop could open up a whole new world of possibilities. 

The M1 hybrid thing was an alternative to making stand alone Desktop apps which is out of reach for me, since I can't handle the certificates coding. Plus the HD icon resolution hasn't been added in Animate for about 10 years. There is also the issue of the desktop app being copies and distributed without DRM..

**Reply by ajwfrost** [reply]
So:
a) we've fixed the issue with 'native' code-signing not picking up the difference between a distribution certificate vs a developer one, which was why the app then failed to install earlier.
b) we've verified that applications that are requesting a landscape layout are appearing in landscape mode.

I think it's all working then ... but there may be some APIs that don't work properly perhaps. If anyone comes across an issue with an API that doesn't work as expected when running an IPA on a Mac, please raise it! Although, there are of course some limitations inherent in this...

**Reply by wasserworks** [reply]
That's potentially great news - I wonder which version of Air fixed the landscape issue. 

Let's ask some testers who have an M1 Mac!

**Reply by FliplineStudios** [reply]
That's great news!  We're in the same boat where we don't have an M1 Mac ourselves yet to check, we had just heard from users with M1 Macs having the right side of the screen covered up in landscape-only apps, and had to disable M1 compatibility across the board for all of our apps in App Store Connect.  (Attached is a screenshot of what users were seeing for reference, and to be clear this was sent in 2020 from an older app, I believe built with AIR 32 still)
![mac_m1_landscapeissue](https://user-images.githubusercontent.com/33292459/150185652-11842b6c-7aaf-48f5-b125-24161cb41a63.png)

@ajwfrost Is that the behavior you were seeing with landscape apps showing up as portrait, and which should now be fixed in an upcoming version?

**Reply by ajwfrost** [reply]
Hi @FliplineStudios  - I don't know for sure if that's the same issue, I thought it was more that the application window itself was appearing in portrait mode rather than landscape. The above screenshot looks more as if there's some confusion between the width and height and you're ending up with it only drawing in a square area...

But, something has changed already (possibly within AIR but I'm thinking more likely with macOS) which means that this issue is no longer happening, so it may be worth re-checking. We have M1 macs in case you want to send us an IPA file to test...

thanks
